rasa/shared/core/flows/flow.py

```diff
@@ -274,6 +276,73 @@ def first_step_in_flow(self) -> Optional[FlowStep]:
             return None
         return self.steps[0]
 
+    @staticmethod
+    def slots_from_steps(steps: List[FlowStep]) -> List[str]:
+        """Return the names of the slots used in the given steps."""
+        result = []
+        for step in steps:
+            if isinstance(step, QuestionFlowStep) and step.question not in result:
+                result.append(step.question)
+        return result
+
+    def slots(self) -> List[str]:
+        """Return the names of the slots used in the flow."""
+        return self.slots_from_steps(self.steps)
+
+    def slots_up_to_step(self, id: str) -> List[str]:
+        """Returns the names of the slots used in this flow up to a step."""
+        step_ids = [step.id for step in self.steps]
+        try:
+            idx = step_ids.index(id)
+        except ValueError:
+            idx = -1
+        steps = self.steps[: idx + 1]
+        return self.slots_from_steps(steps)
+
+    def previously_asked_questions(self, step_id: Text) -> List[QuestionFlowStep]:
```

Comment:
@twerkmeister added this if you need it

Reply:
Just as with my above functions, we might get slightly incorrect results for more complicated graphs involving branching and joining patterns.

Reply:
yes agreed, knowing for sure isn't possible atm as we are missing the information on the path we took - we'd need to record that somewhere